### PR TITLE
WIP, ENH: add POSIX access pattern plot to darshan job summary

### PR DESCRIPTION
**Pull request by shanedsnyder**
See commits for more details.

If things look good, I can add some tests that can at least check some expected `summary_record` counter values returned by `acummulate_records()`.

A few thoughts:
- This is the first plotting routine that takes a dataframe as input (others take a report object). I thought about this a bit, and think it would be good to generalize the plotting routines we provide away from report objects, to allow us to plot data coming from multiple log files. I didn't modify any other plotting routines, but could take a pass at this in another PR if that seems reasonable?
- Related to above, but this new plotting routine is hardcoded to get plot data from the first row of the input dataframe for now. That works fine for the accumulator use case obviously since it's just a single row, but I'm not sure how to best generalize this. Seems like it should be easy? Ultimately, I think it would be nice if you could give the plotting routine a single row (in which case it plots the row values directly) or an entire dataframe (in which case it will need to run the accumulator internally to condense down to a single record). Any thoughts on this?
- Can we convert the derived metrics structure returned by `accumulate_records()` into a more Pythonic object? `summary_record` is a dictionary now (i.e., not a cdata object), so seems that it would be cleaner if both return values are typical Python things, rather than `derived_metrics` being a C-like structure.

### CODE REVIEW

**Comment by shanedsnyder** [comment]
Here's what a new example report looks like for now (e3sm_io_heatmap_only.darshan):
image

So, obvious things to cleanup are to make sure the legend isn't on top of plot data, and probably to add a page break between plots (or otherwise reorg things so the big table isn't jammed next to this plot).

File: darshan-util/pydarshan/darshan/experimental/plots/plot_posix_access_pattern.py
Comment (tylerjereddy) [inline darshan-util/pydarshan/darshan/experimental/plots/plot_posix_access_pattern.py:26]: > This is the first plotting routine that takes a dataframe as input 

This is also a bit confusing for public consumption--it is actually a dictionary with the single-row data frames and other stuff nested inside of it. That will probably need to be crystal clear for end users eventually.

File: darshan-util/pydarshan/darshan/backend/cffi_backend.py
Comment (tylerjereddy) [reply darshan-util/pydarshan/darshan/backend/cffi_backend.py:793]: Finally, for the comment:

> Can we convert the derived metrics structure returned by accumulate_records() into a more Pythonic object? summary_record is a dictionary now (i.e., not a cdata object), so seems that it would be cleaner if both return values are typical Python things, rather than derived_metrics being a C-like structure.

Maybe that should be a separate issue as well--if it substantially reduces lines of code maybe. It is true that I don't like how there is a `struct` than then contains an `array` which has to be manually indexed based on knowledge of the C code to pull out some counter values. Maybe sketch out a design in another issue if you plan to flatten to a custom `class` or maybe even `dataclass` is appropriate... though dunno where it sits on priority side.

**Review by tylerjereddy — COMMENTED** [review]
> probably to add a page break between plots

On my monitor they aren't too close together--it probably depends on screen size and stuff--probably an html/css layout that auto-adapts or something would be good, but fancier stuff there may be a can of worms (darshan reports on the phone, etc...).

image

It is good that the plot looks similar vs. old perl report on same log file:

image

Some thoughts on the new plot:
- the 9 consecutive read counts are only visible because of the value labels added above the bar chart values--not sure if a threshold for log scaling might be warranted (I don't really care too much)--you could probably argue about the merits of emphasizing size differences vs. not noticing a non-zero entry...
- why isn't the total category equal to the sum of consecutive and sequential (i.e., the read values here?)--if the categories aren't mutually exclusive I wonder if that is worth noting

For the other points, I'd prefer to keep the discussion focused and have issues opened to elaborate on the pros/cons:

> I thought about this a bit, and think it would be good to generalize the plotting routines we provide away from report objects, to allow us to plot data coming from multiple log files. I didn't modify any other plotting routines, but could take a pass at this in another PR if that seems reasonable?

I'm perhaps less convinced of the value/priority for now--I think an issue with a clear description of concrete use cases and how the API would work for the current usages is likely in order. I believe some of our plotting routines behave quite differently, so I'm also not entirely convinced this strategy is appropriate in a super broad sense, and I'd also like the plotting functions as simple as possible--avoid doing data analysis inside the plotting function proper where possible.

If it were clear that this reduced the amount of code substantially that might be convincing, though for a infrastructure that is initially "report"/single file focused, it seems likely this would initially grow more complexity to allow generalization, and may bloat plotting functions with shims/analyses that have little to do with plotting proper.

For folks that need functionality this advanced, I think an argument might need to be made that they can't just write their own plotting code? Thinking about the heatmap plotting, my brain melts a bit re: fusing files and DataFrames from multiple sources, and supporting bugs from users related to this kind of thing. Probably other such examples, and cases that are a bit more on the fence. So, my initial thought is--case-by-case basis with strong real-world justification, and try to keep plotting code devoid of analysis complexity and instead accept the final condensed/consensus data structures?

**Comment by shanedsnyder** [comment]
I'll respond more to other comments later, but wanted to start here:

> > I thought about this a bit, and think it would be good to generalize the plotting routines we provide away from report objects, to allow us to plot data coming from multiple log files. I didn't modify any other plotting routines, but could take a pass at this in another PR if that seems reasonable?
> 
> I'm perhaps less convinced of the value/priority for now--I think an issue with a clear description of concrete use cases and how the API would work for the current usages is likely in order. I believe some of our plotting routines behave quite differently, so I'm also not entirely convinced this strategy is appropriate in a super broad sense, and I'd also like the plotting functions as simple as possible--avoid doing data analysis inside the plotting function proper where possible.
> 
> If it were clear that this reduced the amount of code substantially that might be convincing, though for a infrastructure that is initially "report"/single file focused, it seems likely this would initially grow more complexity to allow generalization, and may bloat plotting functions with shims/analyses that have little to do with plotting proper.
> 

Sure, we could document what each existing plot is expecting in terms of data in an issue or something. At a high-level, all they do now is take a report as input and extract all the data they need from that. I think the change I'm suggesting would be to instead have the caller of the plotting routines responsible for actually extracting the data in the format the plots expect and passing in directly, instead of having the plot routines digging around in report objects. Seems like that would actually simplify plotting code at the expense of requiring callers to be more explicit about the data they want plotted.

Of course, the main driver of this is not really code cleanup, it's just to generalize our plotting routines so that they could conceivably plot data that comes from multiple log files. As it stands, that is impossible, which will be a definite problem going forward. We really need this capability to extend Darshan to be useful in context of analyzing workflows (that generate many log files) or for analyzing tons of logs generated at a facility. I think continuing to tie our plotting infrastructure to single logs would be a mistake in that regard.

> For folks that need functionality this advanced, I think an argument might need to be made that they can't just write their own plotting code? Thinking about the heatmap plotting, my brain melts a bit re: fusing files and DataFrames from multiple sources, and supporting bugs from users related to this kind of thing. Probably other such examples, and cases that are a bit more on the fence. So, my initial thought is--case-by-case basis with strong real-world justification, and try to keep plotting code devoid of analysis complexity and instead accept the final condensed/consensus data structures?

The use cases I mention above are definitely of interest to our users. I'm not sure the changes are really that challenging (we'll see, I think the plot in this PR provides some supporting evidence that this could be relatively simple) that we have to punt on this for users. The heatmaps will almost certainly be most involved, but I don't think we have to figure that out all at once. The other plots seem like they would be incredibly straightforward to switch to a dataframe or record based input, which would make them trivial to apply for single- or multi-log use cases. I don't really see the downside other than that it requires a small refactor?

File: darshan-util/pydarshan/darshan/backend/cffi_backend.py
Comment (shanedsnyder) [reply darshan-util/pydarshan/darshan/backend/cffi_backend.py:793]: A summary record is orthogonal to derived metrics, though they are both notionally based on aggregating sets of module records.

_derived metrics_: Calculates a set of statistics describing high-level I/O characteristics of a set of records, including things like total bytes read/written, total time spent doing I/O, bandwidth calculations, etc. Darshan can provide derived metrics for all of its traditional modules, even though they have different counter names, etc. It's just a way to generalize I/O characteristics of different modules into general I/O behavior that is no longer module specific, if that makes sense.

_summary record_: This literally generates an entirely new module record that is the aggregation of a set of given module records. So, it's not calculating derived metrics, it's just creating a new POSIX record that is the aggregation of a set of input POSIX records, for instance. You could use something like this to turn a bunch of per-process records referencing a file into a single aggregate record for that file (giving you total operation counts, access sizes, etc.).

As an aside, I think basically all of the plots in our existing job summary report could simply be updated to use an  summary record for input, which would simplify things a lot, I think (completely ignoring the issues related to whether this aggregation should be in Python or C -- the aggregation is useful and is now hidden behind an API, we could always move it in the future but I don't think it's really relevant here).

**Comment by tylerjereddy** [comment]
Maybe just shift that stuff to an issue yeah--for code that is basically "produce a bar chart" like here, I don't really even see much value in having it public, could basically just have the aggregator public and a small example in its docstring of how you might condense and plot, freeing us from any plotting code maintenance entirely.

**Comment by tylerjereddy** [comment]
And if DataFrames are embraced more globally in the code base, they already have plotting methods attached to them as well, so calling `DataFrame.plot.bar()` and so on may very well get users close enough in many cases, just need some small docs that are tested/maintained with examples.

**Comment by shanedsnyder** [comment]
> why isn't the total category equal to the sum of consecutive and sequential (i.e., the read values here?)--if the categories aren't mutually exclusive I wonder if that is worth noting

I can see how this could be confusing, as I was convinced there was a problem too before thinking about it more. The issue is that the "sequential" count is inclusive of the "consecutive" count (see the definitions in the figure caption). I.e., you have to subtract the consecutive count from the sequential count to get the count of operations that were _only_ sequential. 

I'll have to think a bit more and see if any tweaks to the plot make sense or if we should at least add some clarify text about this.

File: darshan-util/pydarshan/darshan/backend/cffi_backend.py
Comment (shanedsnyder) [reply darshan-util/pydarshan/darshan/backend/cffi_backend.py]: Agreed with this, fixed in [5f6eec0](https://github.com/darshan-hpc/darshan/pull/903/commits/5f6eec03843519c9762007b30fc07dab18f4198e)

File: darshan-util/pydarshan/darshan/cli/summary.py
Comment (shanedsnyder) [reply darshan-util/pydarshan/darshan/cli/summary.py]: Agreed with this, fixed in [5f6eec0](https://github.com/darshan-hpc/darshan/pull/903/commits/5f6eec03843519c9762007b30fc07dab18f4198e)

File: darshan-util/pydarshan/darshan/tests/test_lib_accum.py
Comment (shanedsnyder) [reply darshan-util/pydarshan/darshan/tests/test_lib_accum.py]: Good call, this is done [a449459](https://github.com/darshan-hpc/darshan/pull/903/commits/a449459d147d325d387fbc6ec09044b7b29e3afa)

File: darshan-util/pydarshan/darshan/experimental/plots/plot_posix_access_pattern.py
Comment (shanedsnyder) [reply darshan-util/pydarshan/darshan/experimental/plots/plot_posix_access_pattern.py:20]: I agree with this generally, but just punted to an issue (gh-923) so we could look at handling this for all plots. Maybe it makes sense to do that as we start moving plotting routines out of experimental namespace or something.

File: darshan-util/pydarshan/darshan/experimental/plots/plot_posix_access_pattern.py
Comment (shanedsnyder) [reply darshan-util/pydarshan/darshan/experimental/plots/plot_posix_access_pattern.py:26]: I'm going to open a separate issue to discuss potential reworking of plots to and what might make sense for different routines to take as input, and hopefully we can clarify stuff like this there.

The intention is to allow users to pass in an individual record to this routine to have data from it plotted -- that record could represent a single file or, as in the case of the job summary tool, an accumulated record of all files accessed by the application.

File: darshan-util/pydarshan/darshan/experimental/plots/plot_posix_access_pattern.py
Comment (shanedsnyder) [reply darshan-util/pydarshan/darshan/experimental/plots/plot_posix_access_pattern.py]: I can get on board with this (having caller responsible for inputting only a single record here, potentially using accumulation logic themselves).

I removed the comment for now to avoid confusion, since I'm not even sure it's possible to have a "record" with multiple rows in the dataframe -- I don't think this is anything we need to sanity check for now.

File: darshan-util/pydarshan/darshan/backend/cffi_backend.py
Comment (shanedsnyder) [reply darshan-util/pydarshan/darshan/backend/cffi_backend.py]: > Maybe that should be a separate issue as well--if it substantially reduces lines of code maybe. It is true that I don't like how there is a struct than then contains an array which has to be manually indexed based on knowledge of the C code to pull out some counter values. Maybe sketch out a design in another issue if you plan to flatten to a custom class or maybe even dataclass is appropriate... though dunno where it sits on priority side.

New issue in gh-924

**Comment by shanedsnyder** [comment]
> I'll have to think a bit more and see if any tweaks to the plot make sense or if we should at least add some clarifying text about this.

I ultimately just added a new figure caption to clarify that sequential is inclusive of consecutive. I'm not sure other tweaks will be much more helpful so I think I'm content with that. We could present "sequential" as being strictly sequential (i.e., does not include consecutive operations), but that also will require clarify text so not sure it's any better.

**Comment by shanedsnyder** [comment]
I think I'm all done with edits here and ready for any final review comments before merging.

**Comment by tylerjereddy** [comment]
I'll try to take a final look on April 17th, got swamped with grants and stuff this weekend

**Comment by tylerjereddy** [comment]
Ok, I looked at the diff again and checked a sample log->HTML locally and it seems "ok" and still consistent with the old `perl` report for access pattern plotting.

Since we really do need to wrap up the feature-completeness of the new Python report I'll go ahead and merge.

I think some of the regression testing didn't really happen here re: the values we see in the new plot, but I'll leave that up to the team I guess. As far as I can tell, we've just incremented the number of expected plots, which is at least better than nothing.

**Review by tylerjereddy — APPROVED** [review]

**Pull request merged**
